File: Wire-iOS/Sources/Helpers/syncengine/Message.m
// 
// Wire
// Copyright (C) 2016 Wire Swiss GmbH
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
// 


#import "Message+Private.h"
#import "Message+Formatting.h"
#import "Constants.h"
#import "Settings.h"

@import WireExtensionComponents;

@implementation Message

+ (MessageType)messageType:(id<ZMConversationMessage>)message
{
    if ([self isImageMessage:message]) {
        return MessageTypeImage;
    }
    if ([self isKnockMessage:message]) {
        return MessageTypePing;
    }
    if ([self isTextMessage:message]) {
        if ((message.textMessageData.linkPreview != nil) || ([self linkAttachments:message.textMessageData].count > 0)) {
            return MessageTypeRichMedia;
        }
        return MessageTypeText;
    }
    if ([self isFileTransferMessage:message]) {
        if (message.fileMessageData.isVideo) {
            return MessageTypeVideo;
        }
        if (message.fileMessageData.isAudio) {
            return MessageTypeAudio;
        }
        return MessageTypeFile;
    }
    if ([self isSystemMessage:message]) {
        return MessageTypeSystem;
    }
    if ([self isLocationMessage:message]) {
        return MessageTypeLocation;
    }
    return MessageTypeUnknown;
}

+ (BOOL)isTextMessage:(id<ZMConversationMessage>)message
{
    return message.textMessageData != nil;
}

+ (BOOL)isImageMessage:(id<ZMConversationMessage>)message
{
    return message.imageMessageData != nil;
}

+ (BOOL)isKnockMessage:(id<ZMConversationMessage>)message
{
    return message.knockMessageData != nil;
}

+ (BOOL)isFileTransferMessage:(id<ZMConversationMessage>)message
{
    return message.fileMessageData != nil;
}

+ (BOOL)isVideoMessage:(id<ZMConversationMessage>)message
{
    return [self isFileTransferMessage:message] && [message.fileMessageData isVideo];
}

+ (BOOL)isAudioMessage:(id<ZMConversationMessage>)message
{
    return [self isFileTransferMessage:message] && [message.fileMessageData isAudio];
}

+ (BOOL)isLocationMessage:(id<ZMConversationMessage>)message
{
    return message.locationMessageData != nil;
}

+ (BOOL)isSystemMessage:(id<ZMConversationMessage>)message
{
    return message.systemMessageData != nil;
}

+ (BOOL)isNormalMessage:(id<ZMConversationMessage>)message
{
    return [self isTextMessage:message] || [self isImageMessage:message] || [self isKnockMessage:message] || [self isFileTransferMessage:message] || [self isVideoMessage:message] || [self isAudioMessage:message] || [self isLocationMessage:message] ;
}

+ (BOOL)isConnectionRequestMessage:(id<ZMConversationMessage>)message
{
    if ([self isSystemMessage:message]) {
        return message.systemMessageData.systemMessageType == ZMSystemMessageTypeConnectionRequest;
    }
    return NO;
}

+ (BOOL)isMissedCallMessage:(id<ZMConversationMessage>)message
{
    if ([self isSystemMessage:message]) {
        return message.systemMessageData.systemMessageType == ZMSystemMessageTypeMissedCall;
    }
    return NO;
}

+ (BOOL)isDeletedMessage:(id<ZMConversationMessage>)message
{
    return [Message isSystemMessage:message] && ((ZMSystemMessage *)message).systemMessageType == ZMSystemMessageTypeMessageDeletedForEveryone;
}

+ (NSString *)formattedReceivedDateForMessage:(id<ZMConversationMessage>)message
{
    // Today's date
    NSDate *today = [NSDate new];
    
    NSDate *serverTimestamp = message.serverTimestamp;
    if (! serverTimestamp) {
        serverTimestamp = today;
    }
    
    return [serverTimestamp wr_formattedDate];
}

+ (BOOL)shouldShowTimestamp:(id<ZMConversationMessage>)message
{
    BOOL allowedType =  [Message isTextMessage:message] ||
                        [Message isImageMessage:message] ||
                        [Message isFileTransferMessage:message] ||
                        [Message isKnockMessage:message] ||
                        [Message isLocationMessage:message] ||
                        [Message isDeletedMessage:message];
    
    return allowedType;
}

+ (NSDateFormatter *)shortVersionDateFormatter
{
    static NSDateFormatter *shortVersionDateFormatter = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        shortVersionDateFormatter = [[NSDateFormatter alloc] init];
        [shortVersionDateFormatter setDateStyle:NSDateFormatterShortStyle];
        [shortVersionDateFormatter setTimeStyle:NSDateFormatterNoStyle];
    });
    
    return shortVersionDateFormatter;
}

+ (NSDateFormatter *)longVersionTimeFormatter
{
    static NSDateFormatter *longVersionTimeFormatter = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        longVersionTimeFormatter = [[NSDateFormatter alloc] init];
        [longVersionTimeFormatter setDateStyle:NSDateFormatterNoStyle];
        [longVersionTimeFormatter setTimeStyle:NSDateFormatterShortStyle];
    });
    
    return longVersionTimeFormatter;
}

+ (BOOL)isPresentableAsNotification:(id<ZMConversationMessage>)message
{
    BOOL isChatHeadsDisabled = [[Settings sharedSettings] chatHeadsDisabled];
    BOOL isConversationSilenced = message.conversation.isSilenced;
    BOOL isSenderSelfUser = [message.sender isSelfUser];
    BOOL isMessageUnread = (NSUInteger) message.serverTimestamp.timeIntervalSinceReferenceDate > message.conversation.lastReadMessage.serverTimestamp.timeIntervalSinceReferenceDate;
    // only show a notification chathead if the message is new (recently sent)
    BOOL isTimelyMessage = [[NSDate date] timeIntervalSinceDate:message.serverTimestamp] <= (0.5f);
    BOOL isSystemMessage = [self isSystemMessage:message];

    return ! isChatHeadsDisabled &&
            ! isConversationSilenced &&
            ! isSenderSelfUser &&
            isMessageUnread &&
            isTimelyMessage &&
            ! isSystemMessage;
}

@end

File: wire-ios-data-model/Source/Model/Message/Message.swift
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation

extension ZMConversationMessage {
    /// Returns YES, if the message has text to display.
    /// This also includes linkPreviews or links to soundcloud, youtube or vimeo
    public var isText: Bool {
        textMessageData != nil
    }

    public var isImage: Bool {
        imageMessageData != nil || (fileMessageData != nil && fileMessageData!.v3_isImage)
    }

    public var isKnock: Bool {
        knockMessageData != nil
    }

    /// Returns YES, if the message is a file transfer message
    /// This also includes audio messages and video messages
    public var isFile: Bool {
        fileMessageData != nil && !fileMessageData!.v3_isImage
    }

    public var isPDF: Bool {
        isFile && fileMessageData?.isPDF ?? false
    }

    public var isPass: Bool {
        isFile && fileMessageData!.isPass
    }

    public var isVideo: Bool {
        isFile && fileMessageData!.isVideo
    }

    public var isAudio: Bool {
        isFile && fileMessageData!.isAudio
    }

    public var isLocation: Bool {
        locationMessageData != nil
    }

    public var isSystem: Bool {
        systemMessageData != nil
    }

    public var isNormal: Bool {
        isText
            || isImage
            || isKnock
            || isFile
            || isVideo
            || isAudio
            || isLocation
    }

    public var isConnectionRequest: Bool {
        guard isSystem else {
            return false
        }
        return systemMessageData!.systemMessageType == .connectionRequest
    }

    public var isMissedCall: Bool {
        guard isSystem else {
            return false
        }
        return systemMessageData!.systemMessageType == .missedCall
    }

    public var isDeletion: Bool {
        guard isSystem else {
            return false
        }
        return systemMessageData!.systemMessageType == .messageDeletedForEveryone
    }
}

extension ConversationCompositeMessage {
    public var isComposite: Bool {
        compositeMessageData != nil
    }
}

// MARK: - Message

/// The `ZMConversationMessage` protocol can not be extended in Objective-C,
/// thus this helper class provides access to commonly used properties.
public class Message: NSObject {
    /// Returns YES, if the message has text to display.
    /// This also includes linkPreviews or links to soundcloud, youtube or vimeo
    @objc(isTextMessage:)
    public class func isText(_ message: ZMConversationMessage) -> Bool {
        message.isText
    }

    @objc(isImageMessage:)
    public class func isImage(_ message: ZMConversationMessage) -> Bool {
        message.isImage
    }

    @objc(isKnockMessage:)
    public class func isKnock(_ message: ZMConversationMessage) -> Bool {
        message.isKnock
    }

    /// Returns YES, if the message is a file transfer message
    /// This also includes audio messages and video messages
    @objc(isFileTransferMessage:)
    public class func isFileTransfer(_ message: ZMConversationMessage) -> Bool {
        message.isFile
    }

    @objc(isPDFMessage:)
    public class func isPDF(_ message: ZMConversationMessage) -> Bool {
        message.isPDF
    }

    @objc(isVideoMessage:)
    public class func isVideo(_ message: ZMConversationMessage) -> Bool {
        message.isVideo
    }

    @objc(isAudioMessage:)
    public class func isAudio(_ message: ZMConversationMessage) -> Bool {
        message.isAudio
    }

    @objc(isLocationMessage:)
    public class func isLocation(_ message: ZMConversationMessage) -> Bool {
        message.isLocation
    }

    @objc(isSystemMessage:)
    public class func isSystem(_ message: ZMConversationMessage) -> Bool {
        message.isSystem
    }

    @objc(isNormalMessage:)
    public class func isNormal(_ message: ZMConversationMessage) -> Bool {
        message.isNormal
    }

    @objc(isConnectionRequestMessage:)
    public class func isConnectionRequest(_ message: ZMConversationMessage) -> Bool {
        message.isConnectionRequest
    }

    @objc(isMissedCallMessage:)
    public class func isMissedCall(_ message: ZMConversationMessage) -> Bool {
        message.isMissedCall
    }

    @objc(isDeletedMessage:)
    public class func isDeleted(_ message: ZMConversationMessage) -> Bool {
        message.isDeletion
    }
}
